File: src/game/components/aggression.lua
local constants = require 'game.components.constants'
local collision = require 'game.utils.collision'

local Aggression = {
  _meta = constants.Aggression
}

function Aggression.new(id, world, entity, x, y, width, height, duration)
  width = width or 250 -- width of aggression box
  height = height or 250 -- height of aggression box
  duration = duration or 3 -- lose aggression after duration

  local body = love.physics.newBody(world, x or 0, y or 0, 'kinematic')
  local shape = love.physics.newRectangleShape(width, height)
  local fixture = love.physics.newFixture(body, shape, 1)
  fixture:setSensor(true)
  fixture:setCategory(8)
  fixture:setUserData({entity = entity, aggression = true})

  local aggression = {
    _meta = Aggression._meta,
    id = id,
    fixture = fixture,
    minRange = minRange,
    maxRange = maxRange,
    duration = duration
  }

  function aggression:destroy()
    self.fixture:getBody():destroy()
  end

  return aggression
end

return Aggression

File: client2/src/game/components/aggression.ts
import { Component } from '../../ecs/component'
import { Flag } from './flags'
import { Entity } from '../../ecs/Entity'

export class Aggression extends Component {
  static _id = 'Aggression'
  _id = Aggression._id

  static _flag = Flag.Aggression
  _flag = Flag.Aggression

  fixture: Fixture
  width: number
  height: number
  duration: number

  constructor(
    world: World,
    entity: Entity,
    x: number = 0,
    y: number = 0,
    width: number = 0,
    height: number = 0,
    duration: number = 0
  ) {
    super()

    this.width = width
    this.height = height
    this.duration = duration

    this.fixture = love.physics.newFixture(
      love.physics.newBody(world, x, y, 'kinematic'),
      love.physics.newRectangleShape(width, height),
      1
    )
    this.fixture.setSensor(true)
    this.fixture.setCategory(8)
    this.fixture.setUserData({ entity: entity, aggression: true })
  }

  destroy = () => {
    this.fixture.getBody().destroy()
  }
}